src/main/java/org/hyperledger/fabric/sdk/Endpoint.java

```diff
@@ -61,243 +70,318 @@
+            throw new RuntimeException(
+                format("Properties \"%s\" and \"%s\" cannot both be set", propertyKey1, propertyKey2)
+            );
+        }
+    }
 
-                        logger.trace(format("Endpoint %s Negotiation type: '%s', SSLprovider: '%s'", url, nt, sslp));
-                        SslProvider sslprovider = sslp.equals("openSSL") ? SslProvider.OPENSSL : SslProvider.JDK;
-                        NegotiationType ntype = nt.equals("TLS") ? NegotiationType.TLS : NegotiationType.PLAINTEXT;
+    private static class GrpcsContext {
 
-                        SslContextBuilder clientContextBuilder = getSslContextBuilder(clientCert, clientKey, sslprovider);
-                        SslContext sslContext;
+        private final String cn;
+        private final byte[] pemBytes;
 
-                        logger.trace(format("Endpoint %s  final server pemBytes: %s", url, Hex.encodeHexString(pemBytes)));
+        private GrpcsContext(
+            String cn, byte[] pemBytes) {
+            this.cn = cn;
+            this.pemBytes = pemBytes;
+        }
 
-                        try (InputStream myInputStream = new ByteArrayInputStream(pemBytes)) {
-                            sslContext = clientContextBuilder
-                                    .trustManager(myInputStream)
-                                    .build();
-                        }
+        public Optional<String> getCn() {
+            return Optional.ofNullable(cn);
+        }
 
-                        channelBuilder = NettyChannelBuilder
-                                .forAddress(addr, port)
-                                .sslContext(sslContext)
-                                .negotiationType(ntype);
+        public Optional<byte[]> getPemBytes() {
+            return Optional.ofNullable(pemBytes);
+        }
+    }
 
-                        if (cn != null) {
-                            logger.debug(format("Endpoint %s, using CN overrideAuthority: '%s'", url, cn));
-                            channelBuilder.overrideAuthority(cn);
-                        }
-                        addNettyBuilderProps(channelBuilder, properties);
-                    } catch (SSLException sslex) {
+    private enum ProtocolType {
+        GRPC,
+        GRPCS;
+
+        private static ProtocolType from(String protocol) {
+            if ("grpc".equalsIgnoreCase(protocol)) {
+                return GRPC;
+            }
+            if ("grpcs".equalsIgnoreCase(protocol)) {
+                return GRPCS;
+            }
+            throw new RuntimeException("invalid protocol: " + protocol);
+        }
+    }
+
+    private static Optional<byte[]> getPEMFile(String url, Properties properties) {
+        byte[] pemBytes;
+        try (ByteArrayOutputStream bis = new ByteArrayOutputStream(64000)) {
+            @SuppressWarnings("UseOfPropertiesAsHashtable") byte[] pb = (byte[]) properties.get("pemBytes");
```

Comment:
It might be worth extracting `"pemBytes"` out as a constant as you have done nicely for the other property names